### DISCUSSION TITLE
Instructions how to use this Project with an ESP32S3

### DISCUSSION THREAD

**Discussion by pschatzmann** [adafruit/Adafruit_TinyUSB_Arduino#462] [Q&A] [answered]
I don't quite understand how this project is supported by the ESP32.
The readme of this project is saying that all that is needed is to add an ```#include<Adafruit_TinyUSB.h> ```

But this is not working when using the 'esp32 by Espressif Systems' core version 3.0.5.

```
/tmp/.arduinoIDE-unsaved2024911-2595-1pellbi.ihjm/sketch_oct11a/sketch_oct11a.ino:1:10: fatal error: Adafruit_TinyUSB.h: No such file or directory
    1 | #include <Adafruit_TinyUSB.h>
      |          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
exit status 1

Compilation error: Adafruit_TinyUSB.h: No such file or directory
```

I also tried to install this project via the library mananager: but then the related device C++ classes can not be found.

**Comment by znmeb** [comment]
I have some interesting ESP32 devices I can test this with. In particular, I have an [M5Stack Cardputer](https://shop.m5stack.com/products/m5stack-cardputer-kit-w-m5stamps3), which has a real microphone and speaker on it in addition to the USB. The speaker isn't exactly HiFi but it's handheld game console quality.

**Comment by pschatzmann** [comment, marked as answer]
This was quite a traumatic experience, but I think I finally understand how things are supposed to work for the ESP32S3.

1) Unlike the other platforms, this library needs to be installed via the Arduino __library manager__. It contains tinyusb as the ESP32 core does and discrepancies will lead to linker errors. So if you get linker errors just [file an issue](https://github.com/adafruit/Adafruit_TinyUSB_Arduino/issues/465) as I did!
   This approach is quite unfortunate, because this installation will __break the other supported platforms__!


2) For this library to work, the ESP32 needs the following __USB settings__ in the Tools menu:

 - USB CDC on Boot: __Enabled__
 - USB DFU on Boot: Disabled
 - USB Firmware MSC on Boot: Disabled
 - Upload Mode:  __USB-OTC  CDC (TinyUSB)__
 - USB Mode:  __USB-OTC (TinyUSB)__

   Many thanks to [Andreas Spiess](https://www.youtube.com/watch?v=hJSBTFsOnoA) for his video!

3) It is advisable to set the ESP32 into upload mode by pressing BOOT, then ENABLE, then release ENABLE then release BOOT (at least when you load the sketch the first time)

4) In Linux you can __check the descriptor__ after you deployed your sketch with ``lsusb -d 303a:1001 -v``. So if you have loaded a midi sketch you should see the CDC and the MIDI descriptor!

 5) The __Tiny USB configuration__ is loaded from the tinyusb component which is part of the core. On linux this is in ``~/.arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-632e0c2a/esp32s3/include/arduino_tinyusb/include/tusb_config.h``   which is basically just referring to settings that are defined in sdkconfig.  (``~/.arduino15/packages/esp32/tools/esp32-arduino-libs/idf-release_v5.1-632e0c2a/esp32s3/sdkconfig  ?) 

As a conclusion, you can't extend the functionality with your custom descriptors (e.g. to implement Audio), w/o submitting a pull request to the ESP32 Arduino core and then you need to wait for a new core release!
